### PR TITLE
Add on the fly generation of presets

### PR DESCRIPTION
**Pull request by krvajal**
A similar approach to the one presented on this blog post 
https://aws.amazon.com/blogs/compute/resize-images-on-the-fly-with-amazon-s3-aws-lambda-and-amazon-api-gateway/ was implemented.

This has several benefits. First, we are be more resilient to errors. If for example, a specific preset generation fails, we are gonna be able to generate it on the fly. This depends, of course, on the fact that the original image exists and can be readed. Secondly, if we on the future decide to add more presets, we will be able to generate them for the existing images on the fly.

The redirect rule configuration is defined as part of the CloudFormation stack definition, in the `resources` key of the `serverless.yml` configuration file.

~~One thing that I didn't manage to achive is to specify the CloudFormation resource key dynamically based on the environment, as variable interpolation does not seem to work in the keys, only on the values. If it worked, instead of having~~
```
S3BucketSweepbrightimagesproduction
```
~~I will use instead somethink like~~
```
	S3BucketSweepbrightimages${opt:stage}
```
~~but it is not possible currently.~~

In a subsequent PR https://github.com/sweepbright/serverless-lambda-resize-image/pull/11 I have updated the logical id of the storage bucket.

I have one posible solution to this problem but it requires some level of duplication. Other solution is to write a custom script that does a find and remplace and sets the correct environment variable. I am open to any other suggestions as well.

I want to also express that I am a bit concerned about the fact having this could be used to execute a DoS attack. So further investigation is needed to control the executions of the lambda functions and provide some security mechanisms

### CODE REVIEW

**Comment by xavierbarbosa** [comment]
> I want to also express that I am a bit concerned about the fact having this could be used to execute a DoS attack. So further investigation is needed to control the executions of the lambda functions and provide some security mechanisms

Do we already have some basic validation as UUID, etc? Cloudflare could help with protecting ourselves against a DoS attack.

**Comment by krvajal** [comment]
We are not using Cloudflare

File: serverless.yml [hunk 1]
```diff
@@ -9,22 +9,69 @@ provider:
...
         - 'lambda:InvokeAsync'
       Resource: '*'
+    # allow to attach events to exiting buckets (serverless-plugin-existing-s3 plugin)
+    -  Effect: "Allow"
+       Action:
+         - "s3:PutBucketNotification"
+       Resource:
+         Fn::Join:
+           - ""
+           - - "arn:aws:s3:::${self:provider.environment.ORIGIN_BUCKET}"
```
Comment (ohjimijimijimi) [inline serverless.yml]: this seems to be a typo
```suggestion
           - "arn:aws:s3:::${self:provider.environment.ORIGIN_BUCKET}"
```

File: serverless.yml [hunk 2]
```diff
@@ -9,22 +9,69 @@ provider:
...
   bucket_name: 'sweepbright-images-${opt:stage}'
+
+resources:
+  Outputs:
+    ApiUrl:
+      Description: "The API Gateway URL"
+      Value:
+        Fn::Join:
+          - ""
+          - - Ref: ApiGatewayRestApi
```
Comment (ohjimijimijimi) [inline serverless.yml]: another type?
```suggestion
          - Ref: ApiGatewayRestApi
```

File: src/utils/urlbuilder.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+function getCDNBasePath(stage) {
```
Comment (ohjimijimijimi) [inline src/utils/urlbuilder.js]: This is not really the path. We could call it 
```suggestion
function getCDNDomain(stage) {
```

File: notfound.js [hunk 4]
```diff
@@ -3,38 +3,39 @@ const s3 = require('./src/config/s3');
 const logger = require('./src/utils/logger');
 const transformWithPreset = require('./src/operations/transformWithPreset');
 const savePreset = require('./src/operations/savePreset');
+const UrlBuilder = require('./src/utils/urlbuilder');
 
 module.exports.handler = (evt, ctx, callback) => {
-  const { preset, image_key } = evt.pathParameters;
-
+  const { key } = evt.queryStringParameters;
+  const [group, _, preset, filename] = key.split('/');
```
[the remark this answers is not in the document]
Comment (krvajal) [reply notfound.js]: yes, it is

File: serverless.yml (hunk 1 above)
Comment (krvajal) [reply serverless.yml]: not a typo

File: serverless.yml (hunk 2 above)
Comment (krvajal) [reply serverless.yml]: not a typo

File: src/utils/urlbuilder.js (hunk 3 above)
Comment (krvajal) [reply src/utils/urlbuilder.js]: Fair enough

File: serverless.yml (hunk 1 above)
Comment (ohjimijimijimi) [reply serverless.yml]: why the double dashes then?

Comment (krvajal) [reply serverless.yml]: This is how it looks if you express it in json
```
{ Fn::Join: [
	"",
	["arn:aws:s3:::${self:provider.environment.ORIGIN_BUCKET}"]
  ]
}      
```

File: serverless.yml (hunk 2 above)
Comment (krvajal) [reply serverless.yml]: And for this one the json object will be
```
{ Value: {
        Fn::Join:[ "", 
			["Ref: ApiGatewayRestApi",
			 ".execute-api.${self:provider.region}.", 
			 "Ref: AWS::URLSuffix"
			]
		]
 }
}
```

**Pull request merged**